### DISCUSSION TITLE
It is about time

### DISCUSSION THREAD

**Discussion by mtoy-googly-moogly** [malloydata/malloy#458] [What's Next: Language Proposals]
# Unit based time operations in Malloy

### Units

The basic list of units, and the first thing to notice is that while timestamp has millisecond resolution, there is no millisecond unit in Malloy.

* `second`
* `minute`
* `hour`
* `day`
* `week`
* `month`
* `quater`
* `year`

The three basic Malloy gestures dealing with time units are

### Truncation
*  _timeExpr_`.`_unitName_

### Extraction
* _unitName_`(`_timeExpr_`)`

The extraction gesture also recognizes:

* `day_of_week`
* `day_of_year`

### Measurement
* _unitName_`(`_timeExpr_ `to` _timeExpr_`)`

### Plurals

Because the expressions read like unit is describing the result, not the operation, and the result is a number, it seemed natural to allow the plural of the unit in the time gesture.

```
---- truncate
someTime.hour
someTime.hours
---- extract
hour(someTime)
hours(someTime)
---- measure
hour(someTime to anotherTime)
hours(someTime to anotherTime)
```

## What SHOULD we do?

I have a series of questions

1. Do we have the right syntax for the three gestures
1. Do the plurals make sense
1. `day` and `date` is a little confusing

### Different gesture syntaxes

Outside of Malloy the function and dot gestures have a general meaning.

The dot (`.`) gesture carries the connotation of accessing an internal piece of a larger whole, or invoking a method on an object.

The function (`unitName()`) gesture suggests a graph y=f(x). Or to say it another way it is a mapping from one domain to another.

It isn't clear to me that our current usage map to the general understanding of these gestures. In particular, it bothers me that if I showed 100 people `someVar.hour` and `hour(someVar)` and asked them, "which is telling me the hour part and which is truncating to an hour?", I suspect the majority would guess the opposite of what the current meaning is.

I was talked out of this earlier, but I kind of have to come back and try one more time.

### A proposal

The extraction gesture is most similar to "modulus", which in most programming languages is the `%` operator

If `someTime` was the timestamp `2001-02-03 04:05:06` ...

```
someTime % seconds => 6
someTime % year => 2001
someTime % day => 3
someTime % day_of_week => some number from 1 to 7
someTime % day_of_year => some number from 1 to 366
```

Measurements would also use `%`

```
(someTime to anotherTime) % day // days between two times
```

Which leaves us the question of how you express truncation, is dot the correct operator for truncation? If dot has any natural meaning, it would be "access the named part of the whole", and so it feels odd that dot means "do a truncation"

I've always felt like the function gesture was kind of weird, but is it weirder than inventing a new operator which means truncate?

Because internally we treat truncated times like ranges, I always wanted truncation to use `::` implying that the type of this object is changing,

```
someTime::hour => 2001-02-03 00:00:00 with granularity 1 hour
```

I feel like this makes sense because

```
someTime::hour + 1 => 2001-02-03 01:00:00
```

I especially like this because it makes these two gestures naturally aliases for each other, except with the clear connotation that one is a date, and one is a timestamp with zeroes on clock-units.

```
someTime::day
someTime::date
```

And I would eliminate the plurals, because of the way the unit names are seen as a unit specification of the operation, not as description of the result.

### New syntax sample

Here are our three original samples, in the proposed new syntax

```
---- truncate
someTime::hour
---- extract
someTime % hour
---- measure
(someTime to anotherTime) % hour
```

**Comment by anikaks** [comment]
I'll share my variety of thoughts about time...
- I agree that `day` and `date` is confusing. My instinct that I am still getting caught on is that it should be `created_at.date`, not `created_at.day`.
- I don't mind the dot syntax at all, though I do see what you're saying about it implying something a little different from what it's doing, this hasn't confused me in practice (yet). I find it very pleasingly easy to type and read.
- I'd love to dig more into what it is you don't like about `unit()`. I am actually quite fond of this option; it's concise, and SQL uses functions for all these things (granted it's `DATE_TRUNC()`, `EXTRACT()`, and `DATE_DIFF()` and they take timeframe arguments, but anyway...). My feeling is we gain familiarity and intuitiveness by using function syntax. 
- One other thing I want to talk through is that I actually find the differentiation between extraction and truncation a bit confusing. I would be _very_ happy to see us settle on a solution where truncation and extraction look more alike. I know this is not a popular opinion since they do produce different types, but to me it feels _much_ more intuitive than thinking through whether I want a truncation or an extraction and which syntax is which each time. To make this happen (and perhaps anyway), I'd be in favor of typing out the slight longer `week_of_year`, `day_of_week`, `hour_of_day`, etc for all extractions, as I am often confusing the extractions and which extraction `day` is. 
- Getting into the specific syntax you proposed, I can't say I'm _wild_ about the `::` as it makes me think of casting which this isn't precisely, nor the `%` which makes me think of a wildcard.

**Reply by mtoy-googly-moogly** [reply]
SQL is so broken, and time is so integral to data, that don't mind being different. Maybe we should be showing the world what it looks like for time to be truly supported in a language instead of trying to look like SQL.

In malloy, every time object has two attributes, a base type and a granularity. In the expression`foo <  now.year +1 or foo > now.day - 1` we see two granularities (day and year) . If we view that pair of base type/granularity as the "type" then maybe it makes sense to use the `::` operator to change the granularity.

I think the `x % day` and `x::day` is also addressing the problem of "if i see `day(x)` and `x.day`, which is truncation and which is extraction", maybe it doesn't do it well, but it is absolutely trying to it better.

**Reply by mtoy-googly-moogly** [reply]
I guess another choice would be to provide sql-like functions just without the maddening need to get the types right `trunc(x, day)` `extract(x, day)`, `duration(x to y, days)`

do not love that, but i might like it more than what we have now

**Reply by mtoy-googly-moogly** [reply]
and ... hmm ...

`x % month.day` => 1-31
`x % year.day` => 1-366
`x % week.day` => 1-7

???

**Reply by mtoy-googly-moogly** [reply]
I did forget that in SQL, `%` is a wildcard for LIKE. That might be a money wrench in the `MOD operator` idea.

**Reply by anikaks** [reply]
the `month.day` `year.day` concept is _very_ intriguing!
